incubator/hnc/pkg/reconcilers/hnc_config.go

```diff
@@ -362,21 +383,88 @@ func (r *ConfigReconciler) validateSingletonName(ctx context.Context, nm string)
+	for _, ts := range r.Forest.GetTypeSyncers() {
+		if r.activeGVKs[ts.GetGVK()] && ts.GetMode() != api.Ignore {
+			r.addNumPropagatedObjects(ts.GetGVK(), ts.GetNumPropagatedObjects(), inst)
+		}
+	}
+	sort.Slice(inst.Status.Types, func(i, j int) bool {
+		return inst.Status.Types[i].APIVersion < inst.Status.Types[j].APIVersion
+	})
+}
+
+// addNumPropagatedObjects adds the NumPropagatedObjects field for a given GVK in the status.
+func (r *ConfigReconciler) addNumPropagatedObjects(gvk schema.GroupVersionKind, num int,
+	inst *api.HNCConfiguration) {
+	apiVersion, kind := gvk.ToAPIVersionAndKind()
+	n := int32(num)
+	inst.Status.Types = append(inst.Status.Types, api.TypeSynchronizationStatus{
+		APIVersion:           apiVersion,
+		Kind:                 kind,
+		NumPropagatedObjects: &n,
+	})
+}
+
+// enqueueSingleton enqueues the `config` singleton to trigger the reconciliation
+// of the singleton for a given reason . This occurs in a goroutine so the
 // caller doesn't block; since the reconciler is never garbage-collected,
 // this is safe.
-func (r *ConfigReconciler) forceInitialReconcile(log logr.Logger, reason string) {
+func (r *ConfigReconciler) enqueueSingleton(log logr.Logger, reason string) {
 	go func() {
 		log.Info("Enqueuing for reconciliation", "reason", reason)
 		// The watch handler doesn't care about anything except the metadata.
 		inst := &api.HNCConfiguration{}
 		inst.ObjectMeta.Name = api.HNCConfigSingleton
-		r.Igniter <- event.GenericEvent{Meta: inst}
+		r.Trigger <- event.GenericEvent{Meta: inst}
 	}()
 }
 
+func (r *ConfigReconciler) getShouldReconcile() bool {
+	r.shouldReconcileLock.Lock()
+	defer r.shouldReconcileLock.Unlock()
+
+	return r.shouldReconcile
+}
+
+func (r *ConfigReconciler) unsetShouldReconcile() {
+	r.shouldReconcileLock.Lock()
+	defer r.shouldReconcileLock.Unlock()
+
+	r.shouldReconcile = false
+}
+
+// periodicTrigger periodically checks if the `config` singleton needs to be reconciled and
+// enqueues the `config` singleton for reconciliation, if needed.
+// Object reconcilers signal the config reconciler to reconcile when the status needs to
+// be updated. The config reconciler only reconciles periodically so that the reconciliation
+// won't be triggered too frequently.
+func (r *ConfigReconciler) periodicTrigger() {
+	// run forever
+	for {
+		time.Sleep(checkPeriod)
+		if r.getShouldReconcile() == false {
+			continue
+		}
+		r.enqueueSingleton(r.Log, "Syncing NumPropagatedObjects in the status")
+		r.unsetShouldReconcile()
+	}
+}
+
+// SyncNumPropagatedObjects will be called by object reconcilers to signal config
+// reconciler to reconcile when an object is reconciled successfully and the status of
+// the `config` object might need to be updated.
+func (r *ConfigReconciler) SyncNumPropagatedObjects(log logr.Logger) {
+	r.shouldReconcileLock.Lock()
+	defer r.shouldReconcileLock.Unlock()
+
+	log.Info("Signalling config reconciler for reconciliation.")
```

Comment:
This will be very noisy - I'd suggest log.V(1).Info

Reply:
Sure! Done.